utils/include/detray/simulation/event_generator/random_track_generator.hpp

```diff
@@ -16,48 +16,74 @@
-#include <cmath>
+#include <array>
 #include <random>
 
 namespace detray {
 
-/// Wrapper for random number generatrion for the @c random_track_generator
+/// Wrapper for CPU random number generatrion for the @c random_track_generator
 template <typename scalar_t = scalar,
           typename distribution_t = std::uniform_real_distribution<scalar_t>,
-          typename generator_t = std::random_device,
           typename engine_t = std::mt19937_64>
 struct random_numbers {
 
-    random_numbers(random_numbers&& other) : engine(std::move(other.engine)) {}
-
-    generator_t gen;
-    engine_t engine;
-
-    template <
-        typename T = generator_t,
-        std::enable_if_t<std::is_same_v<T, std::random_device>, bool> = true>
-    random_numbers() : gen{}, engine{gen()} {}
-
-    template <typename T = generator_t,
-              std::enable_if_t<std::is_same_v<T, std::seed_seq>, bool> = true>
-    random_numbers() : gen{42u}, engine{gen} {}
-
-    template <typename T = distribution_t,
-              std::enable_if_t<
-                  std::is_same_v<T, std::uniform_real_distribution<scalar_t>>,
-                  bool> = true>
-    DETRAY_HOST auto operator()(const scalar_t min, const scalar_t max) {
-        return distribution_t(min, max)(engine);
+    using distribution_type = distribution_t;
+    using engine_type = engine_t;
+    using seed_type = typename engine_t::result_type;
+
+    std::seed_seq m_seeds;
+    engine_t m_engine;
+
+    /// Default seed
+    DETRAY_HOST
+    random_numbers()
+        : m_seeds{random_numbers::default_seed()}, m_engine{m_seeds} {}
+
+    /// Different seed @param s for every instance
+    DETRAY_HOST
+    random_numbers(seed_type s) : m_seeds{s}, m_engine{m_seeds} {}
+
+    /// More entropy in seeds from collection @param s
+    DETRAY_HOST
+    random_numbers(const std::vector<seed_type>& s)
+        : m_seeds{s.begin(), s.end()}, m_engine{m_seeds} {}
+
+    /// Copy constructor
+    DETRAY_HOST
+    random_numbers(random_numbers&& other)
+        : m_engine(std::move(other.m_engine)) {}
+
+    /// Generate random numbers in a given range
+    DETRAY_HOST auto operator()(const std::array<scalar_t, 2> range = {
+                                    -std::numeric_limits<scalar_t>::max(),
+                                    std::numeric_limits<scalar_t>::max()}) {
+        const scalar_t min{range[0]}, max{range[1]};
+        assert(min <= max);
+
+        // Uniform
+        if constexpr (std::is_same_v<
+                          distribution_t,
+                          std::uniform_real_distribution<scalar_t>>) {
+
+            return distribution_t(min, max)(m_engine);
+
+            // Normal
+        } else if constexpr (std::is_same_v<
+                                 distribution_t,
+                                 std::normal_distribution<scalar_t>>) {
+
+            scalar_t mu{min + 0.5f * (max - min)};
+            return distribution_t(mu, 0.5f / 3.0f * (max - min))(m_engine);
```

Comment:
This is what I meant. Just change `range` to `param`
```suggestion
   return distribution_t(param[0], param[1])(m_engine);
```

Reply:
This is done in l.81 and not what this method should do

Reply:
OK. Then I would suggest removing the function  `normal` and use this one only. The suggestion stills goes the same. 

The reason is that `normal` should not outstand like this because this is not supposed to be used for uniform distribution right?

Reply:
And imagine setting -1 ,1  for normal. then mean will be 0 and stddev will be 2/6. 
Why should be stddev set like this? 1/6 wouldn't be better? or how about 1/1000? It is not possible set a good standard

Reply:
But then I would have to do the same calculation for the mean and standard deviation that is in this method every time before I call it. That would lead to code duplication...

I need this method to scale the mean and standard deviation such, that a high percentage of the Gaussian distribution lies between ```min``` and ```max```. So, if 100 and 180  is passed, say x% of all of the numbers drawn from the Gaussian should lie between 100 and 180 and the mean should be at 140.

Reply:
x% should be such, that clamping the numbers drawn from this to 100 and 180 only removes a small part of the distributions tail

Reply:
The ```normal``` method is also used when the ```random_number``` struct is set to uniform distribution, namely for the vertex smearing

Reply:
So I am suggesting to have what is in `normal` in `operator()` as it is useful

Reply:
No problem! I think we might want to try this soon to simulate jet-like objects, even though I am not sure how well this fits....

Reply:
> x% should be such,

So how will you set x exactly? 
I understand you are trying to give a unified interface for gaussian and uniform distribution -- but every single person will have different ideas in `what is the good value of x`. And I don't agree with 2/6.

Reply:
I don't know, I am open to suggestions. We can add this to the config struct of the random track generator?

Reply:
Maybe we can put it in multiples of sigma environment?

Reply:
SImulating jets.. 😃 only if you have parton distribution available

Reply:
> I don't know, I am open to suggestions. We can add this to the config struct of the random track generator?

I would just take parameter directly instead of adding config struct.

The best way would be separating current track generators into three: (random_uniform_track_generator, random_gauss_track_generator, fixed_uniform_track_generator)

Reply:
:D It does not have to be very realistic, I don't believe. I am also not sure I understood what exactly we should produce correctly, whether it should be jets or just a bundle of particles in one Region of Interest, idk. Let's discuss in the meeting tomorrow how best to proceed with this